### PR TITLE
Adding plots for each dataset in core.py

### PR DESCRIPTION
**Pull request by tanmayy24**
This PR introduces a suite of visualization and analysis tools for audio datasets, enhancing understanding and interpretation of dataset characteristics. It includes class and subclass distribution visualizations, interactive waveform and mel spectrogram plotting, and a comprehensive statistical analysis of the dataset. These enhancements offer insights into class balance, dataset composition, and audio characteristics, aiding in more informed data processing and model training decisions.

### CODE REVIEW

File: soundata/core.py [hunk 1]
```diff
@@ -361,6 +377,459 @@ def validate(self, verbose=True):
         )
         return missing_files, invalid_checksums
 
+    @lru_cache(maxsize=None)  # Setting maxsize to None for an unbounded cache
+    def compute_clip_statistics(self):
+        durations = [
+            len(self.clip(c_id).audio[0]) / self.clip(c_id).audio[1]
+            for c_id in tqdm(
+                list(self._index["clips"].keys())[:100], desc="Calculating durations"
```
Comment (magdalenafuentes) [inline soundata/core.py]: Why is this topped to 100?

File: soundata/core.py [hunk 2]
```diff
@@ -6,7 +6,23 @@
 import types
 from typing import Any
 
+from functools import lru_cache
```
Comment (magdalenafuentes) [inline soundata/core.py]: Could you group imports that are for the figure with a comment? e.g. `# Dataset plots necessary imports`. There are a lot of them...

File: soundata/core.py [hunk 3]
```diff
@@ -361,6 +377,459 @@ def validate(self, verbose=True):
         )
         return missing_files, invalid_checksums
 
+    @lru_cache(maxsize=None)  # Setting maxsize to None for an unbounded cache
+    def compute_clip_statistics(self):
+        durations = [
+            len(self.clip(c_id).audio[0]) / self.clip(c_id).audio[1]
```
Comment (magdalenafuentes) [inline soundata/core.py]: Should we add a line checking if the attribute `audio` is present in the dataset just in case?

File: soundata/core.py [hunk 4]
```diff
@@ -361,6 +377,459 @@ def validate(self, verbose=True):
...
+            for c_id in tqdm(
+                list(self._index["clips"].keys())[:100], desc="Calculating durations"
+            )
+        ]
+
+        # Calculate statistics
+        total_duration = sum(durations)
+        mean_duration = np.mean(durations)
+        median_duration = np.median(durations)
```
Comment (magdalenafuentes) [inline soundata/core.py]: IIRC there was a problem with one of this that was showing the wrong estimate, is that working now?

File: soundata/core.py [hunk 5]
```diff
@@ -361,6 +377,459 @@ def validate(self, verbose=True):
...
+                total_duration_unit = "minutes"
+        else:
+            if total_duration > 120:
+                total_duration /= 60
+                total_duration_unit = "minutes"
+            else:
+                total_duration_unit = "seconds"
+
+        # Define the base colors
+        base_colors = ["#404040", "#126782", "#C9C9C9"]
```
Comment (magdalenafuentes) [inline soundata/core.py]: Do we only need three colors for all datasets?

**Review by magdalenafuentes — CHANGES_REQUESTED** [review]
Hey @tanmayy24 this is a great addition, thanks! See some comments below, I haven't finished looking at it yet. 

One general comment I have is that it's a bit change to incorporate directly into the `core.py`, and I think it will be cleaner if we move this to a module like it's done for `annotations` and others. We could call that `display.py`. What do you think?

File: soundata/core.py (hunk 1 above)
Comment (tanmayy24) [reply soundata/core.py:385]: Fixed

File: soundata/core.py (hunk 5 above)
Comment (tanmayy24) [reply soundata/core.py:442]: Colors to resemble the soundata template

**Review by magdalenafuentes — APPROVED** [review]

**Pull request merged**